# 01. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
# create path

path = path = r'/Users/martyna/Documents/CF Data Analysis/Data Immersion/Achievement 6/COVID Analysis 8.22'

In [3]:
# import covid cases data

df_covid = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','US-COVID-Levels-County.csv'), index_col = False)

In [4]:
df_covid.head(10)

county  county_fips  \
0                                American Samoa        60000   
1                                          Guam        66000   
2  Commonwealth of the Northern Mariana Islands        69000   
3                  United States Virgin Islands        78000   
4                                American Samoa        60000   
5                                          Guam        66000   
6  Commonwealth of the Northern Mariana Islands        69000   
7                  United States Virgin Islands        78000   
8                                American Samoa        60000   
9                                          Guam        66000   

                                          state  county_population  \
0                                American Samoa            47392.0   
1                                          Guam           168489.0   
2  Commonwealth of the Northern Mariana Islands            51851.0   
3                  United States Virgin Islands           106290.0   
4                                American Samoa            47392.0   
5                                          Guam           168489.0   
6  Commonwealth of the Northern Mariana Islands            51851.0   
7                  United States Virgin Islands           106290.0   
8                                American Samoa            47392.0   
9                                          Guam           168489.0   

   health_service_area_number                           health_service_area  \
0                       901.0                                American Samoa   
1                       902.0                                          Guam   
2                       903.0  Commonwealth of the Northern Mariana Islands   
3                       905.0                  United States Virgin Islands   
4                       901.0                                American Samoa   
5                       902.0                                          Guam   
6                       903.0  Commonwealth of the Northern Mariana Islands   
7                       905.0                  United States Virgin Islands   
8                       901.0                                American Samoa   
9                       902.0                                          Guam   

   health_service_area_population  covid_inpatient_bed_utilization  \
0                         47392.0                              0.0   
1                        168489.0                             15.2   
2                         51851.0                              7.2   
3                        106290.0                              2.4   
4                         47392.0                              0.9   
5                        168489.0                             13.7   
6                         51851.0                              8.3   
7                        106290.0                              1.4   
8                         47392.0                              1.0   
9                        168489.0                             12.3   

   covid_hospital_admissions_per_100k  covid_cases_per_100k  \
0                                 2.1                156.14   
1                                38.6                833.88   
2                                 9.6               1267.09   
3                                 0.9                 78.09   
4                                19.0                434.67   
5                                77.2                450.47   
6                                 3.9                779.16   
7                                 0.0                 31.99   
8                                 6.3                901.00   
9                                83.1                385.78   

  covid-19_community_level date_updated  
0                      Low       3/3/22  
1                     High       3/3/22  
2                   Medium       3/3/22  
3                      Low       3/3/22  
4                   

# 03. Data wrangling

In [5]:
# check data shape

df_covid.shape

(80600, 12)

##  i. Dropping columns

In [6]:
# dropping columns for irrelevant/redundant information

df_covid_clean = df_covid.drop(['county_fips', 'health_service_area_number', 'health_service_area', 'health_service_area_population'], axis = 1)

In [7]:
df_covid_clean.shape

(80600, 8)

In [8]:
df_covid_clean.head(10)

county  \
0                                American Samoa   
1                                          Guam   
2  Commonwealth of the Northern Mariana Islands   
3                  United States Virgin Islands   
4                                American Samoa   
5                                          Guam   
6  Commonwealth of the Northern Mariana Islands   
7                  United States Virgin Islands   
8                                American Samoa   
9                                          Guam   

                                          state  county_population  \
0                                American Samoa            47392.0   
1                                          Guam           168489.0   
2  Commonwealth of the Northern Mariana Islands            51851.0   
3                  United States Virgin Islands           106290.0   
4                                American Samoa            47392.0   
5                                          Guam           168489.0   
6  Commonwealth of the Northern Mariana Islands            51851.0   
7                  United States Virgin Islands           106290.0   
8                                American Samoa            47392.0   
9                                          Guam           168489.0   

   covid_inpatient_bed_utilization  covid_hospital_admissions_per_100k  \
0                              0.0                                 2.1   
1                             15.2                                38.6   
2                              7.2                                 9.6   
3                              2.4                                 0.9   
4                              0.9                                19.0   
5                             13.7                                77.2   
6                              8.3                                 3.9   
7                              1.4                                 0.0   
8                              1.0                                 6.3   
9                             12.3                                83.1   

   covid_cases_per_100k covid-19_community_level date_updated  
0                156.14                      Low       3/3/22  
1                833.88                     High       3/3/22  
2               1267.09                   Medium       3/3/22  
3                 78.09                      Low       3/3/22  
4                434.67                     High      3/10/22  
5                450.47                     High      3/10/22  
6                779.16                   Medium      3/10/22  
7                 31.99                      Low      3/10/22  
8                901.00                   Medium      3/17/22  
9                385.78                     High      3/17/22

## ii. Data types

In [9]:
# checking data types  

df_covid_clean.dtypes

county                                 object
state                                  object
county_population                     float64
covid_inpatient_bed_utilization       float64
covid_hospital_admissions_per_100k    float64
covid_cases_per_100k                  float64
covid-19_community_level               object
date_updated                           object
dtype: object

In [10]:
# checking for mixed data types

for col in df_covid_clean.columns.tolist():
  weird = (df_covid_clean[[col]].applymap(type) != df_covid_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_covid_clean[weird]) > 0:
    print (col)

covid-19_community_level


In [11]:
# changing data type to string

df_covid_clean['covid-19_community_level'] = df_covid_clean['covid-19_community_level'].astype('str')

In [12]:
df_covid_clean['covid-19_community_level'].dtype

dtype('O')

## iii. Duplicates

In [13]:
# finding duplicate values

df_covid_clean[df_covid_clean.duplicated()]

Empty DataFrame
Columns: [county, state, county_population, covid_inpatient_bed_utilization, covid_hospital_admissions_per_100k, covid_cases_per_100k, covid-19_community_level, date_updated]
Index: []

no duplicate values found

## iv. Missing data

In [14]:
df_covid_clean.isnull().sum()

county                                  0
state                                   0
county_population                       5
covid_inpatient_bed_utilization       149
covid_hospital_admissions_per_100k     34
covid_cases_per_100k                    0
covid-19_community_level                0
date_updated                            0
dtype: int64

In [15]:
df_nan = df_covid_clean[df_covid_clean['county_population'].isnull()== True]

In [16]:
df_nan

county                         state  \
9571            Chugach Census Area                        Alaska   
9678       Copper River Census Area                        Alaska   
18098      Copper River Census Area                        Alaska   
18099           Chugach Census Area                        Alaska   
37541  United States Virgin Islands  United States Virgin Islands   

       county_population  covid_inpatient_bed_utilization  \
9571                 NaN                              NaN   
9678                 NaN                              NaN   
18098                NaN                              NaN   
18099                NaN                              NaN   
37541                NaN                              7.3   

       covid_hospital_admissions_per_100k  covid_cases_per_100k  \
9571                                  NaN                669.05   
9678                                  NaN                102.77   
18098                                 NaN                 34.26   
18099                                 NaN                 62.24   
37541                                 NaN                  0.00   

      covid-19_community_level date_updated  
9571                       nan      3/17/22  
9678                       nan      3/17/22  
18098                      nan      3/31/22  
18099                      nan      3/31/22  
37541                      Low      5/12/22

Missing 5 values in county population column. However there is still data in the covid_cases_per_100k column. Although there are missing values in the population, we still have information on COVID-19 rates to use for the analysis. Will not remove missing values, cannot change to NA as this will change the data type to mixed and affect analysis.

In [17]:
df_nan2 = df_covid_clean[df_covid_clean['covid_inpatient_bed_utilization'].isnull()== True]

In [18]:
df_nan2

county  \
127                                  Jackson County   
2541                                    Cass County   
2609                                Harrison County   
2666                                  Marion County   
2841                                   Henry County   
...                                             ...   
76027                                 Monroe County   
77773                                 Monroe County   
79335                                Audrain County   
79356                             Montgomery County   
79792  Commonwealth of the Northern Mariana Islands   

                                              state  county_population  \
127                                        Arkansas            16719.0   
2541                                          Texas            30026.0   
2609                                          Texas            66553.0   
2666                                          Texas             9854.0   
2841                                       Virginia            50557.0   
...                                             ...                ...   
76027                                      Missouri             8644.0   
77773                                      Missouri             8644.0   
79335                                      Missouri            25388.0   
79356                                      Missouri            11551.0   
79792  Commonwealth of the Northern Mariana Islands            56889.0   

       covid_inpatient_bed_utilization  covid_hospital_admissions_per_100k  \
127                                NaN                                 NaN   
2541                               NaN                                 NaN   
2609                               NaN                                 NaN   
2666                               NaN                                 NaN   
2841                               NaN                                 NaN   
...                                ...                                 ...   
76027                              NaN                                 0.0   
77773                              NaN                                 0.0   
79335                              NaN                                 0.0   
79356                              NaN                                 0.0   
79792                              NaN                                 NaN   

       covid_cases_per_100k covid-19_community_level date_updated  
127                  263.17                      nan      2/24/22  
2541                 276.43                   Medium      2/24/22  
2609                 220.88                   Medium      2/24/22  
2666                 253.70                   Medium      2/24/22  
2841                 458.89                      nan      2/24/22  
...                     ...                      ...          ...  
76027                150.39                      Low      7/28/22  
77773                370.20                   Medium      8/11/22  
79335                193.00                      Low      8/11/22  
79356                225.09                   Medium      8/11/22  
79792                258.43                      nan      8/11/22  

[149 rows x 8 columns]

There are 149 missing values from the covid_inpatient_bed_utilization column. This is small portion of the dataset, and other columns will be useful for analysis. Will not remove the rows, cannot change to NA as this will change the data type to mixed and affect analysis.

In [19]:
df_nan3 = df_covid_clean[df_covid_clean['covid_hospital_admissions_per_100k'].isnull()== True]

In [20]:
df_nan3

county  \
127                                  Jackson County   
2541                                    Cass County   
2609                                Harrison County   
2666                                  Marion County   
2841                                   Henry County   
2866                                 Patrick County   
2916                              Martinsville City   
9571                            Chugach Census Area   
9678                       Copper River Census Area   
18098                      Copper River Census Area   
18099                           Chugach Census Area   
25613                               Harrison County   
27262                                   Cass County   
37540  Commonwealth of the Northern Mariana Islands   
37541                  United States Virgin Islands   
45984                                American Samoa   
47797                                 Marion County   
54821                                   Cass County   
55636                                 Marion County   
56445                               Harrison County   
58263                                 Marion County   
59024                                   Cass County   
59861                               Harrison County   
61672                                 Hardin County   
62456                                McNairy County   
62507                                 Marion County   
63324                               Harrison County   
63458  Commonwealth of the Northern Mariana Islands   
64087                                   Cass County   
65069  Commonwealth of the Northern Mariana Islands   
68290  Commonwealth of the Northern Mariana Islands   
73998                                American Samoa   
74827  Commonwealth of the Northern Mariana Islands   
79792  Commonwealth of the Northern Mariana Islands   

                                              state  county_population  \
127                                        Arkansas            16719.0   
2541                                          Texas            30026.0   
2609                                          Texas            66553.0   
2666                                          Texas             9854.0   
2841                                       Virginia            50557.0   
2866                                       Virginia            17608.0   
2916                                       Virginia            12554.0   
9571                                         Alaska                NaN   
9678                                         Alaska                NaN   
18098                                        Alaska                NaN   
18099                                        Alaska                NaN   
25613                                         Texas            66553.0   
27262                                         Texas            30026.0   
37540  Commonwealth of the Northern Mariana Islands            51851.0   
37541                  United States Virgin Islands                NaN   
45984                                American Samoa            55461.0   
47797                                         Texas             9854.0   
54821                                         Texas            30026.0   
55636                                         Texas             9854.0   
56445                                         Texas            66553.0   
58263                                         Texas             9854.0   
59024                                         Texas            30026.0   
59861                                         Texas            66553.0   
61672                                     Tennessee            25652.0   
62456                                     Tennessee            25694.0   
62507                                         Texas             9854.0   
63324                                         Texas            66553.0   
63458  Commonwealth of the Northern Mariana Isl

There are 34 missing values in covid_hospital_admissions_per_100k. Once again this is a small portion of the data set, other columns are useful for analysis, will not remove the rows. Cannot replace with NA as this will change the data type to mixed and affect analysis.

## v. Descriptive statistics

In [21]:
df_covid_clean.describe()

county_population  covid_inpatient_bed_utilization  \
count       8.059500e+04                     80451.000000   
mean        1.029197e+05                         3.114527   
std         3.293673e+05                         2.805408   
min         8.600000e+01                         0.000000   
25%         1.113100e+04                         1.100000   
50%         2.611800e+04                         2.500000   
75%         6.721500e+04                         4.300000   
max         1.003911e+07                        36.000000   

       covid_hospital_admissions_per_100k  covid_cases_per_100k  
count                        80566.000000          80600.000000  
mean                             7.299820            145.618164  
std                              6.930791            202.768115  
min                              0.000000              0.000000  
25%                              2.500000             36.547500  
50%                              5.800000            101.200000  
75%                             10.200000            199.020000  
max                            171.200000          13017.750000

The statistics seem reasonable for the dataset. No changes

In [22]:
df_covid_clean.shape

(80600, 8)

# 04. Importing extra smaller data sets

## i. Vaccination data set

In [23]:
# import state vaccination data

df_vax = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','covid19-vaccinations.csv'), index_col = False)

In [24]:
df_vax.head()

State/Territory/Federal Entity  Total Doses Distributed  \
0                  United States                799444535   
1                         Alaska                  1732765   
2                        Alabama                 10551040   
3                       Arkansas                  6893480   
4                 American Samoa                   123210   

   Doses Distributed per 100K  18+ Doses Distributed per 100K  \
0                    240790.0                        309603.0   
1                    236864.0                        314156.0   
2                    215187.0                        276576.0   
3                    228427.0                        297434.0   
4                    259981.0                        393202.0   

   Total Doses Administered by State where Administered  \
0                                          606162842      
1                                            1197172      
2                                            6525635      
3                                            4435639      
4                                             113181      

   Doses Administered per 100k by State where Administered  \
0                                           182574.0         
1                                           163650.0         
2                                           133090.0         
3                                           146982.0         
4                                           238819.0         

   18+ Doses Administered by State where Administered  \
0                                          546209579    
1                                            1066236    
2                                            6075461    
3                                            4038887    
4                                              84267    

   18+ Doses Administered per 100K by State where Administered  \
0                                           211532.0             
1                                           193312.0             
2                                           159257.0             
3                                           174267.0             
4                                           268923.0             

   People with at least One Dose by State of Residence  \
0                                          261981618     
1                                             521311     
2                                            3127008     
3                                            2051724     
4                                              45758     

   Percent of Total Pop with at least One Dose by State of Residence  ...  \
0                                               78.9                  ...   
1                                               71.3                  ...   
2                                               63.8                  ...   
3                                               68.0                  ...   
4                                               95.0                  ...   

   People with a Second Booster Dose  People 50+ with a Second Booster Dose  \
0                         22299699.0                               21064657   
1                                NaN                                  39808   
2                                NaN                                 161345   
3                                NaN                                 139618   
4                                NaN                                   1700   

   Percent of People 50+ with First Booster Dose who Received Second Booster Dose  \
0                                               32.7                                
1                                               33.2                                
2                                               23.0                                
3                                               29.5                                
4                                               20.6            

In [25]:
df_vax.shape

(64, 92)

In [26]:
# change column name to match and merge

df_vax.rename(columns = {'State/Territory/Federal Entity' : 'state'}, inplace = True)

In [27]:
df_vax.head()

state  Total Doses Distributed  Doses Distributed per 100K  \
0   United States                799444535                    240790.0   
1          Alaska                  1732765                    236864.0   
2         Alabama                 10551040                    215187.0   
3        Arkansas                  6893480                    228427.0   
4  American Samoa                   123210                    259981.0   

   18+ Doses Distributed per 100K  \
0                        309603.0   
1                        314156.0   
2                        276576.0   
3                        297434.0   
4                        393202.0   

   Total Doses Administered by State where Administered  \
0                                          606162842      
1                                            1197172      
2                                            6525635      
3                                            4435639      
4                                             113181      

   Doses Administered per 100k by State where Administered  \
0                                           182574.0         
1                                           163650.0         
2                                           133090.0         
3                                           146982.0         
4                                           238819.0         

   18+ Doses Administered by State where Administered  \
0                                          546209579    
1                                            1066236    
2                                            6075461    
3                                            4038887    
4                                              84267    

   18+ Doses Administered per 100K by State where Administered  \
0                                           211532.0             
1                                           193312.0             
2                                           159257.0             
3                                           174267.0             
4                                           268923.0             

   People with at least One Dose by State of Residence  \
0                                          261981618     
1                                             521311     
2                                            3127008     
3                                            2051724     
4                                              45758     

   Percent of Total Pop with at least One Dose by State of Residence  ...  \
0                                               78.9                  ...   
1                                               71.3                  ...   
2                                               63.8                  ...   
3                                               68.0                  ...   
4                                               95.0                  ...   

   People with a Second Booster Dose  People 50+ with a Second Booster Dose  \
0                         22299699.0                               21064657   
1                                NaN                                  39808   
2                                NaN                                 161345   
3                                NaN                                 139618   
4                                NaN                                   1700   

   Percent of People 50+ with First Booster Dose who Received Second Booster Dose  \
0                                               32.7                                
1                                               33.2                                
2                                               23.0                                
3                                               29.5                                
4                                               20.6                                

   People 65+ with a Second Booster Dose  \
0                               14

In [28]:
pd.set_option('display.max_columns', None)

In [29]:
df_vax.head()

state  Total Doses Distributed  Doses Distributed per 100K  \
0   United States                799444535                    240790.0   
1          Alaska                  1732765                    236864.0   
2         Alabama                 10551040                    215187.0   
3        Arkansas                  6893480                    228427.0   
4  American Samoa                   123210                    259981.0   

   18+ Doses Distributed per 100K  \
0                        309603.0   
1                        314156.0   
2                        276576.0   
3                        297434.0   
4                        393202.0   

   Total Doses Administered by State where Administered  \
0                                          606162842      
1                                            1197172      
2                                            6525635      
3                                            4435639      
4                                             113181      

   Doses Administered per 100k by State where Administered  \
0                                           182574.0         
1                                           163650.0         
2                                           133090.0         
3                                           146982.0         
4                                           238819.0         

   18+ Doses Administered by State where Administered  \
0                                          546209579    
1                                            1066236    
2                                            6075461    
3                                            4038887    
4                                              84267    

   18+ Doses Administered per 100K by State where Administered  \
0                                           211532.0             
1                                           193312.0             
2                                           159257.0             
3                                           174267.0             
4                                           268923.0             

   People with at least One Dose by State of Residence  \
0                                          261981618     
1                                             521311     
2                                            3127008     
3                                            2051724     
4                                              45758     

   Percent of Total Pop with at least One Dose by State of Residence  \
0                                               78.9                   
1                                               71.3                   
2                                               63.8                   
3                                               68.0                   
4                                               95.0                   

   People 18+ with at least One Dose by State of Residence  \
0                                          232341244         
1                                             457954         
2                                            2882393         
3                                            1840630         
4                                              32569         

   Percent of 18+ Pop with at least One Dose by State of Residence  \
0                                               90.0                 
1                                               83.0                 
2                                               75.6                 
3                                               79.4                 
4                                               95.0                 

   People Fully Vaccinated by State of Residence  \
0                                      223457170   
1                                         462264   
2                                        2550007   
3                                        1672829   
4                                       

In [30]:
# create subset with necessary columns

df_vax_small = df_vax[['state', 'Total Doses Distributed', 'Doses Distributed per 100K', 'Percent of Total Pop with at least One Dose by State of Residence', 'People Fully Vaccinated by State of Residence', 'Percent of Total Pop Fully Vaccinated by State of Residence']]

In [31]:
df_vax_small.head()

state  Total Doses Distributed  Doses Distributed per 100K  \
0   United States                799444535                    240790.0   
1          Alaska                  1732765                    236864.0   
2         Alabama                 10551040                    215187.0   
3        Arkansas                  6893480                    228427.0   
4  American Samoa                   123210                    259981.0   

   Percent of Total Pop with at least One Dose by State of Residence  \
0                                               78.9                   
1                                               71.3                   
2                                               63.8                   
3                                               68.0                   
4                                               95.0                   

   People Fully Vaccinated by State of Residence  \
0                                      223457170   
1                                         462264   
2                                        2550007   
3                                        1672829   
4                                          42156   

   Percent of Total Pop Fully Vaccinated by State of Residence  
0                                               67.3            
1                                               63.2            
2                                               52.0            
3                                               55.4            
4                                               89.0

In [32]:
df_vax_small.shape

(64, 6)

In [33]:
# change column names

df_vax_small.rename(columns = {'Total Doses Distributed' : 'total_doses_given'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/181491846.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vax_small.rename(columns = {'Total Doses Distributed' : 'total_doses_given'}, inplace = True)


In [34]:
df_vax_small.rename(columns = {'Doses Distributed per 100K' : 'doses_per_100k'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/4034901527.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vax_small.rename(columns = {'Doses Distributed per 100K' : 'doses_per_100k'}, inplace = True)


In [35]:
df_vax_small.rename(columns = {'Percent of Total Pop with at least One Dose by State of Residence' : 'percent_with_at_least_one_dose'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/3240365430.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vax_small.rename(columns = {'Percent of Total Pop with at least One Dose by State of Residence' : 'percent_with_at_least_one_dose'}, inplace = True)


In [36]:
df_vax_small.rename(columns = {'People Fully Vaccinated by State of Residence' : 'people_fully_vaccinated'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/3057078437.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vax_small.rename(columns = {'People Fully Vaccinated by State of Residence' : 'people_fully_vaccinated'}, inplace = True)


In [37]:
df_vax_small.rename(columns = {'Percent of Total Pop Fully Vaccinated by State of Residence' : 'percent_fully_vaccinated'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/27034103.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vax_small.rename(columns = {'Percent of Total Pop Fully Vaccinated by State of Residence' : 'percent_fully_vaccinated'}, inplace = True)


In [38]:
df_vax_small.head()

state  total_doses_given  doses_per_100k  \
0   United States          799444535        240790.0   
1          Alaska            1732765        236864.0   
2         Alabama           10551040        215187.0   
3        Arkansas            6893480        228427.0   
4  American Samoa             123210        259981.0   

   percent_with_at_least_one_dose  people_fully_vaccinated  \
0                            78.9                223457170   
1                            71.3                   462264   
2                            63.8                  2550007   
3                            68.0                  1672829   
4                            95.0                    42156   

   percent_fully_vaccinated  
0                      67.3  
1                      63.2  
2                      52.0  
3                      55.4  
4                      89.0

In [39]:
# check data types

df_vax_small.dtypes

state                              object
total_doses_given                   int64
doses_per_100k                    float64
percent_with_at_least_one_dose    float64
people_fully_vaccinated             int64
percent_fully_vaccinated          float64
dtype: object

In [40]:
# checking for mixed data types

for col in df_vax_small.columns.tolist():
  weird = (df_vax_small[[col]].applymap(type) != df_vax_small[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_vax_small[weird]) > 0:
    print (col)

no columns printed, no mixed type columns

In [41]:
# check duplicates 

df_vax_small[df_vax_small.duplicated()]

Empty DataFrame
Columns: [state, total_doses_given, doses_per_100k, percent_with_at_least_one_dose, people_fully_vaccinated, percent_fully_vaccinated]
Index: []

no duplicates

In [42]:
# check missing values 

df_vax_small.isnull().sum()

state                             0
total_doses_given                 0
doses_per_100k                    3
percent_with_at_least_one_dose    3
people_fully_vaccinated           0
percent_fully_vaccinated          3
dtype: int64

In [43]:
df_vax_nan = df_vax_small[df_vax_small['doses_per_100k'].isnull()== True]

In [44]:
df_vax_nan

state  total_doses_given  doses_per_100k  \
6   Bureau of Prisons             367230             NaN   
11    Dept of Defense            6931410             NaN   
57    Veterans Health            8569020             NaN   

    percent_with_at_least_one_dose  people_fully_vaccinated  \
6                              NaN                   137500   
11                             NaN                  3385029   
57                             NaN                  2978547   

    percent_fully_vaccinated  
6                        NaN  
11                       NaN  
57                       NaN

Total 9 missing values from 3 rows, other information in rows useful for analysis, will not remove. Will not replace to avoid creating mixed data type.

In [45]:
df_vax_small.shape

(64, 6)

## ii. COVID-19 cases and deaths data set

In [46]:
# import state COVID-19 cases and deaths data

df_cases = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','US-cases-deaths-tesing.csv'), index_col = False)

In [47]:
df_cases.head()

State/Territory  Total Cases  Confirmed Cases  Probable Cases  \
0         Alabama    1451998.0        1034836.0        417162.0   
1          Alaska     275542.0              0.0             0.0   
2  American Samoa       8042.0              0.0             0.0   
3         Arizona    2224964.0        1953161.0        271803.0   
4        Arkansas     914619.0              0.0             0.0   

   Cases in Last 7 Days  Case Rate per 100000  Total Deaths  Confirmed Deaths  \
0               13181.0               29613.0       20027.0           15336.0   
1                1868.0               37666.0        1296.0               NaN   
2                  90.0               16969.0          33.0               NaN   
3               13501.0               30568.0       30901.0           26882.0   
4                6229.0               30307.0       11804.0               NaN   

   Probable Deaths  Deaths in Last 7 Days  Death Rate per 100000  \
0           4691.0                    4.0                  408.0   
1              NaN                   21.0                  177.0   
2              NaN                    0.0                   69.0   
3           4019.0                   59.0                  424.0   
4              NaN                   59.0                  391.0   

   7-Day Cases Rate per 100000  7-Day Death Rate per 100000 Total % Positive  \
0                        268.8                          0.1         10-14.9%   
1                        255.4                          2.9           5-7.9%   
2                        189.9                          0.0              NaN   
3                        185.5                          0.8         10-14.9%   
4                        206.4                          2.0         10-14.9%   

  % Positive Last 30 Days % Positive Last 7 Days  # Tests per 100K  \
0                   25% +               20-24.9%       163657.7188   
1                20-24.9%               20-24.9%       577816.8330   
2                     NaN                    NaN               NaN   
3                   25% +               20-24.9%       207185.5705   
4                15-19.9%               15-19.9%       164707.1033   

   Total # Tests  # Tests per 100K Last 7 Days  Total # Tests Last 7 Days  \
0      8054467.0                        726.79                    35769.0   
1      4224754.0                       1532.09                    11202.0   
2            NaN                           NaN                        NaN   
3     15376072.0                        531.52                    39446.0   
4      4991485.0                        659.66                    19991.0   

   # Tests per 100K Last 30 Days  Total # Tests Last 30 Days  New Cases  \
0                        3337.15                    164239.0     1936.0   
1                        6698.83                     48979.0        0.0   
2                            NaN                         NaN        0.0   
3                        2341.40                    173765.0        0.0   
4                        3168.00                     96007.0     1265.0   

   New Deaths  
0         0.0  
1         0.0  
2         0.0  
3         0.0  
4        10.0

In [48]:
# make subset with relevant data

df_cases_small = df_cases[['State/Territory', 'Total Cases', 'Case Rate per 100000', 'Total Deaths', 'Death Rate per 100000', 'Total % Positive']]

In [49]:
df_cases_small.head()

State/Territory  Total Cases  Case Rate per 100000  Total Deaths  \
0         Alabama    1451998.0               29613.0       20027.0   
1          Alaska     275542.0               37666.0        1296.0   
2  American Samoa       8042.0               16969.0          33.0   
3         Arizona    2224964.0               30568.0       30901.0   
4        Arkansas     914619.0               30307.0       11804.0   

   Death Rate per 100000 Total % Positive  
0                  408.0         10-14.9%  
1                  177.0           5-7.9%  
2                   69.0              NaN  
3                  424.0         10-14.9%  
4                  391.0         10-14.9%

In [50]:
# change column names to match and merge

df_cases_small.rename(columns = {'State/Territory' : 'state'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/3547192698.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cases_small.rename(columns = {'State/Territory' : 'state'}, inplace = True)


In [51]:
df_cases_small.rename(columns = {'Total Cases' : 'total_cases'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/63915509.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cases_small.rename(columns = {'Total Cases' : 'total_cases'}, inplace = True)


In [52]:
df_cases_small.rename(columns = {'Case Rate per 100000' : 'case_rate_per_100k'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/123259783.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cases_small.rename(columns = {'Case Rate per 100000' : 'case_rate_per_100k'}, inplace = True)


In [53]:
df_cases_small.rename(columns = {'Total Deaths' : 'total_deaths'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/3009287234.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cases_small.rename(columns = {'Total Deaths' : 'total_deaths'}, inplace = True)


In [54]:
df_cases_small.rename(columns = {'Death Rate per 100000' : 'death_rate_per_100k'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/4197362344.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cases_small.rename(columns = {'Death Rate per 100000' : 'death_rate_per_100k'}, inplace = True)


In [55]:
df_cases_small.rename(columns = {'Total % Positive' : 'total_percent_positive'}, inplace = True)

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/831896674.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cases_small.rename(columns = {'Total % Positive' : 'total_percent_positive'}, inplace = True)


In [56]:
df_cases_small.head()

state  total_cases  case_rate_per_100k  total_deaths  \
0         Alabama    1451998.0             29613.0       20027.0   
1          Alaska     275542.0             37666.0        1296.0   
2  American Samoa       8042.0             16969.0          33.0   
3         Arizona    2224964.0             30568.0       30901.0   
4        Arkansas     914619.0             30307.0       11804.0   

   death_rate_per_100k total_percent_positive  
0                408.0               10-14.9%  
1                177.0                 5-7.9%  
2                 69.0                    NaN  
3                424.0               10-14.9%  
4                391.0               10-14.9%

In [57]:
# check data types

df_cases_small.dtypes

state                      object
total_cases               float64
case_rate_per_100k        float64
total_deaths              float64
death_rate_per_100k       float64
total_percent_positive     object
dtype: object

In [58]:
# checking for mixed data types

for col in df_cases_small.columns.tolist():
  weird = (df_cases_small[[col]].applymap(type) != df_cases_small[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cases_small[weird]) > 0:
    print (col)

total_percent_positive


In [59]:
# changing data type to string

df_cases_small['total_percent_positive'] = df_cases_small['total_percent_positive'].astype('str')

/var/folders/gd/zq96wcpj6d54521g77cwf7sh0000gn/T/ipykernel_25107/3153078406.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cases_small['total_percent_positive'] = df_cases_small['total_percent_positive'].astype('str')


In [60]:
df_cases_small['total_percent_positive'].dtype

dtype('O')

In [61]:
# check duplicates 

df_cases_small[df_cases_small.duplicated()]

Empty DataFrame
Columns: [state, total_cases, case_rate_per_100k, total_deaths, death_rate_per_100k, total_percent_positive]
Index: []

no duplicates

In [62]:
# check missing values 

df_cases_small.isnull().sum()

state                     0
total_cases               1
case_rate_per_100k        1
total_deaths              1
death_rate_per_100k       1
total_percent_positive    0
dtype: int64

In [63]:
df_cases_nan = df_cases_small[df_cases_small['total_cases'].isnull()== True]

In [64]:
df_cases_nan

state  total_cases  case_rate_per_100k  \
10  Federated States of Micronesia          NaN                 NaN   

    total_deaths  death_rate_per_100k total_percent_positive  
10           NaN                  NaN                    nan

No values in entire row, will remove.

In [65]:
df_cases_small = df_cases_small.drop(labels=10, axis=0)

In [66]:
df_cases_small.head(15)

state  total_cases  case_rate_per_100k  total_deaths  \
0                Alabama    1451998.0             29613.0       20027.0   
1                 Alaska     275542.0             37666.0        1296.0   
2         American Samoa       8042.0             16969.0          33.0   
3                Arizona    2224964.0             30568.0       30901.0   
4               Arkansas     914619.0             30307.0       11804.0   
5             California   10844550.0             27446.0       93193.0   
6               Colorado    1612701.0             28004.0       13073.0   
7            Connecticut     863132.0             24209.0       11164.0   
8               Delaware     297879.0             30590.0        3028.0   
9   District of Columbia     164373.0             23291.0        1374.0   
11               Florida    6909415.0             32170.0       78556.0   
12               Georgia    2803517.0             26405.0       39319.0   
13                  Guam      56625.0             33608.0         385.0   
14                Hawaii     324519.0             22920.0        1594.0   
15                 Idaho     483255.0             27042.0        5059.0   

    death_rate_per_100k total_percent_positive  
0                 408.0               10-14.9%  
1                 177.0                 5-7.9%  
2                  69.0                    nan  
3                 424.0               10-14.9%  
4                 391.0               10-14.9%  
5                 235.0                    nan  
6                 227.0                 8-9.9%  
7                 313.0                 5-7.9%  
8                 310.0                 8-9.9%  
9                 194.0                 3-4.9%  
11                365.0               10-14.9%  
12                370.0               10-14.9%  
13                228.0               10-14.9%  
14                112.0                 5-7.9%  
15                283.0               15-19.9%

# 05. Merging data sets

In [67]:
# merge covid county data set with vaccine data set

df_covid_merge = df_covid_clean.merge(df_vax_small, on ='state', indicator = True)

In [68]:
df_covid_merge.head()

county           state  county_population  \
0  American Samoa  American Samoa            47392.0   
1  American Samoa  American Samoa            47392.0   
2  American Samoa  American Samoa            47392.0   
3  American Samoa  American Samoa            47392.0   
4  American Samoa  American Samoa            47392.0   

   covid_inpatient_bed_utilization  covid_hospital_admissions_per_100k  \
0                              0.0                                 2.1   
1                              0.9                                19.0   
2                              1.0                                 6.3   
3                             10.4                                 0.0   
4                              4.1                                80.2   

   covid_cases_per_100k covid-19_community_level date_updated  \
0                156.14                      Low       3/3/22   
1                434.67                     High      3/10/22   
2                901.00                   Medium      3/17/22   
3               1093.01                     High      4/14/22   
4                267.98                     High      4/21/22   

   total_doses_given  doses_per_100k  percent_with_at_least_one_dose  \
0             123210        259981.0                            95.0   
1             123210        259981.0                            95.0   
2             123210        259981.0                            95.0   
3             123210        259981.0                            95.0   
4             123210        259981.0                            95.0   

   people_fully_vaccinated  percent_fully_vaccinated _merge  
0                    42156                      89.0   both  
1                    42156                      89.0   both  
2                    42156                      89.0   both  
3                    42156                      89.0   both  
4                    42156                      89.0   both

In [69]:
df_covid_merge.shape

(79002, 14)

In [70]:
# drop _merge column

df_covid_merge = df_covid_merge.drop(['_merge'], axis=1)

In [71]:
# merge covid case and deaths data set with merged data set

df_covid_merged = df_covid_merge.merge(df_cases_small, on ='state', indicator = True)

In [72]:
df_covid_merged.head(20)

county           state  county_population  \
0   American Samoa  American Samoa            47392.0   
1   American Samoa  American Samoa            47392.0   
2   American Samoa  American Samoa            47392.0   
3   American Samoa  American Samoa            47392.0   
4   American Samoa  American Samoa            47392.0   
5   American Samoa  American Samoa            47392.0   
6   American Samoa  American Samoa            47392.0   
7   American Samoa  American Samoa            47392.0   
8   American Samoa  American Samoa            47392.0   
9   American Samoa  American Samoa            47392.0   
10  American Samoa  American Samoa            47392.0   
11  American Samoa  American Samoa            55461.0   
12  American Samoa  American Samoa            55461.0   
13  American Samoa  American Samoa            55461.0   
14  American Samoa  American Samoa            55461.0   
15  American Samoa  American Samoa            55461.0   
16  American Samoa  American Samoa            55461.0   
17  American Samoa  American Samoa            55461.0   
18  American Samoa  American Samoa            55461.0   
19  American Samoa  American Samoa            55461.0   

    covid_inpatient_bed_utilization  covid_hospital_admissions_per_100k  \
0                               0.0                                 2.1   
1                               0.9                                19.0   
2                               1.0                                 6.3   
3                              10.4                                 0.0   
4                               4.1                                80.2   
5                               4.7                                10.6   
6                               2.1                                 2.1   
7                               1.8                                 0.0   
8                               2.8                                19.0   
9                               1.6                                25.3   
10                             15.2                                48.5   
11                              0.2                                 0.0   
12                              NaN                                 NaN   
13                              0.7                                 5.4   
14                              1.1                                12.6   
15                              0.3                                 3.6   
16                              1.0                                 5.4   
17                              3.4                                10.8   
18                              1.1                                 7.2   
19                              3.2                                43.3   

    covid_cases_per_100k covid-19_community_level date_updated  \
0                 156.14                      Low       3/3/22   
1                 434.67                     High      3/10/22   
2                 901.00                   Medium      3/17/22   
3                1093.01                     High      4/14/22   
4                 267.98                     High      4/21/22   
5                 324.95                     High      4/28/22   
6                 118.16                      Low       5/5/22   
7                 248.99                   Medium      5/12/22   
8                3143.99                     High      3/31/22   
9                2561.61                     high      3/24/22   
10               3238.94                     High       4/7/22   
11                156.14                      Low      5/19/22   
12                 94.95                      nan      5/26/22   
13                173.02                      Low       6/2/22   
14                151.92                   Medium      6/30/22   
15                  0.00                      Low       6/9/22   
16                164.58                      Low      6/16/22   
17                  0.00                   Medium 

In [73]:
df_covid_merged.shape

(79002, 19)

In [74]:
df_covid_merged = df_covid_merged.drop(['_merge'], axis=1)

In [75]:
df_covid_merged.head(5)

county           state  county_population  \
0  American Samoa  American Samoa            47392.0   
1  American Samoa  American Samoa            47392.0   
2  American Samoa  American Samoa            47392.0   
3  American Samoa  American Samoa            47392.0   
4  American Samoa  American Samoa            47392.0   

   covid_inpatient_bed_utilization  covid_hospital_admissions_per_100k  \
0                              0.0                                 2.1   
1                              0.9                                19.0   
2                              1.0                                 6.3   
3                             10.4                                 0.0   
4                              4.1                                80.2   

   covid_cases_per_100k covid-19_community_level date_updated  \
0                156.14                      Low       3/3/22   
1                434.67                     High      3/10/22   
2                901.00                   Medium      3/17/22   
3               1093.01                     High      4/14/22   
4                267.98                     High      4/21/22   

   total_doses_given  doses_per_100k  percent_with_at_least_one_dose  \
0             123210        259981.0                            95.0   
1             123210        259981.0                            95.0   
2             123210        259981.0                            95.0   
3             123210        259981.0                            95.0   
4             123210        259981.0                            95.0   

   people_fully_vaccinated  percent_fully_vaccinated  total_cases  \
0                    42156                      89.0       8042.0   
1                    42156                      89.0       8042.0   
2                    42156                      89.0       8042.0   
3                    42156                      89.0       8042.0   
4                    42156                      89.0       8042.0   

   case_rate_per_100k  total_deaths  death_rate_per_100k  \
0             16969.0          33.0                 69.0   
1             16969.0          33.0                 69.0   
2             16969.0          33.0                 69.0   
3             16969.0          33.0                 69.0   
4             16969.0          33.0                 69.0   

  total_percent_positive  
0                    nan  
1                    nan  
2                    nan  
3                    nan  
4                    nan

In [76]:
df_covid_merged.shape

(79002, 18)

# 06. Export data 

In [77]:
# Exporting as csv file

df_covid_merged.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'covid_merged.csv'))